indexer/indexer.go

```diff
@@ -30,149 +30,228 @@ import (
+	out := &Indexer{
+		log:             log,
+		metricsRegistry: metrics.NewRegistry(),
+		shutdown:        shutdown,
+	}
+	if err := out.initFromConfig(ctx, cfg); err != nil {
+		return nil, errors.Join(err, out.Stop(ctx))
+	}
+	return out, nil
+}
+
+func (ix *Indexer) Start(ctx context.Context) error {
+	// If any of these services has a critical failure,
+	// the service can request a shutdown, while providing the error cause.
+	if err := ix.L1ETL.Start(ix.shutdown); err != nil {
+		return fmt.Errorf("failed to start L1 ETL: %w", err)
+	}
+	if err := ix.L2ETL.Start(ix.shutdown); err != nil {
+		return fmt.Errorf("failed to start L2 ETL: %w", err)
+	}
+	if err := ix.BridgeProcessor.Start(ix.shutdown); err != nil {
+		return fmt.Errorf("failed to start bridge processor: %w", err)
+	}
+	return nil
+}
+
+func (ix *Indexer) Stop(ctx context.Context) error {
+	var result error
+
+	if ix.L1ETL != nil {
+		if err := ix.L1ETL.Close(); err != nil {
+			result = errors.Join(result, fmt.Errorf("failed to close L1 ETL: %w", err))
+		}
+	}
+
+	if ix.L2ETL != nil {
+		if err := ix.L2ETL.Close(); err != nil {
+			result = errors.Join(result, fmt.Errorf("failed to close L2 ETL: %w", err))
+		}
+	}
+
+	if ix.BridgeProcessor != nil {
+		if err := ix.BridgeProcessor.Close(); err != nil {
+			result = errors.Join(result, fmt.Errorf("failed to close bridge processor: %w", err))
+		}
+	}
+
+	// Now that the ETLs are closed, we can stop the RPC clients
+	if ix.l1Client != nil {
+		ix.l1Client.Close()
+	}
+	if ix.l2Client != nil {
+		ix.l2Client.Close()
+	}
+
+	if ix.apiServer != nil {
+		if err := ix.apiServer.Close(); err != nil {
+			result = errors.Join(result, fmt.Errorf("failed to close indexer API server: %w", err))
+		}
+	}
+
+	// DB connection can be closed last, after all its potential users have shut down
+	if ix.DB != nil {
+		if err := ix.DB.Close(); err != nil {
+			result = errors.Join(result, fmt.Errorf("failed to close DB: %w", err))
+		}
+	}
+
+	if ix.metricsServer != nil {
+		if err := ix.metricsServer.Close(); err != nil {
+			result = errors.Join(result, fmt.Errorf("failed to close metrics server: %w", err))
+		}
+	}
+
+	ix.stopped.Store(true)
+
+	ix.log.Info("indexer stopped")
+
+	return result
+}
+
+func (ix *Indexer) Stopped() bool {
+	return ix.stopped.Load()
+}
+
+func (ix *Indexer) initFromConfig(ctx context.Context, cfg *config.Config) error {
+	if err := ix.initRPCClients(ctx, cfg.RPCs); err != nil {
+		return fmt.Errorf("failed to start RPC clients: %w", err)
+	}
+	if err := ix.initDB(ctx, cfg.DB); err != nil {
+		return fmt.Errorf("failed to init DB: %w", err)
+	}
+	if err := ix.initL1ETL(cfg.Chain); err != nil {
+		return fmt.Errorf("failed to init L1 ETL: %w", err)
+	}
+	if err := ix.initL2ETL(cfg.Chain); err != nil {
+		return fmt.Errorf("failed to init L2 ETL: %w", err)
+	}
+	if err := ix.initBridgeProcessor(cfg.Chain); err != nil {
```

Comment:
same here about having the routines put into `Start`. Like the earlier comment I could go either way though